nntrainer/graph/network_graph.cpp

```diff
@@ -430,12 +430,15 @@ void NetworkGraph::setBatchSize(unsigned int batch_size) {
 
   for (auto iter = cbegin(); iter != cend(); iter++) {
     (*iter)->setBatch(batch_size);
-    const InitLayerContext &init_context = (*iter)->getInitContext();
-    // resize tensors spec
-    for (auto const &ts : init_context.getTensorsSpec()) {
-      tensor_manager->setBatchSize(std::get<3>(ts), batch_size);
-      tensor_manager->setBatchSize(std::get<3>(ts) + Var_Grad::grad_suffix,
-                                   batch_size);
+    if ((*iter)->isRunContextAvailable()) {
+      const RunLayerContext &context = (*iter)->getRunContext();
```

Comment:
When is the RunContext available? after finalizing the layer? and i wonder when this setBatchSize called? and I also wonder the difference between first setBatch and the other. 
The first one is only for the input and output .. and the the other is for the tensor which is used during calculation?

Reply:
> When is the RunContext available? after finalizing the layer?

Yes, its right. In this PR in last commit, `isRunContextAvailable` is renamed to `isFinalized()`. 

> and i wonder when this setBatchSize called? and I also wonder the difference between first setBatch and the other.

`setBatchSize()` is called with every call to train/validation/inference. Although the memory is not allocated at that moment, but the size/shapes are set (in case of inference, memory can be allocated as well). In case of first setBatch, memory wont be allocated, and this code will always run. With subsequent calls, if the batch size is same, this function can be a no-op.

> The first one is only for the input and output .. and the the other is for the tensor which is used during calculation?

L441 sets the batch size for the individual tensors by name - this covers the tensors requested by the layers.
L449 sets batch size for the tensors whose behaviors are known - weights, gradients, inputs, outputs, and derivatives.